tests/pub/test_pub_amis.py

```diff
@@ -0,0 +1,300 @@
+            origin="/fake/path/aws/",
+            build=None,
+            build_info=None,
+            signing_key=None,
+            release=AmiRelease(
+                product="SAP",
+                date="20230109",
+                arch="x86_64",
+                respin=0,
+                version="8.4.0",
+                base_product="RHEL",
+                base_version=None,
+                variant="BaseOS",
+                type=None,
+            ),
+            type="hourly",
+            region="us-east-1",
+            virtualization="hvm",
+            volume="gp2",
+            root_device="/dev/sda1",
+            description="Provided by Red Hat, Inc.",
+            sriov_net_support="simple",
+            ena_support=True,
+            billing_codes=AmiBillingCodes(name="Hourly2", codes=["bp-fake"]),
+            image_id="ami-fake-200-A",
+        ),
+        AmiPushItem(
+            name="RHEL-SAP-8.4.0_HVM-20230109-x86_64-0-Hourly2-GP2",
+            state="PENDING",
+            src="/fake/path/aws/me-central-1-hourly/AWS_IMAGES/rhel-sap-ec2-8.4-20220802.sp.10.x86_64.raw",
+            dest=["me-central-1-hourly"],
+            md5sum=None,
+            sha256sum=None,
+            origin="/fake/path/aws/",
+            build=None,
+            build_info=None,
+            signing_key=None,
+            release=AmiRelease(
+                product="SAP",
+                date="20230109",
+                arch="x86_64",
+                respin=0,
+                version="8.4.0",
+                base_product="RHEL",
+                base_version=None,
+                variant="BaseOS",
+                type=None,
+            ),
+            type="hourly",
+            region="me-central-1",
+            virtualization="hvm",
+            volume="gp2",
+            root_device="/dev/sda1",
+            description="Provided by Red Hat, Inc.",
+            sriov_net_support="simple",
+            ena_support=True,
+            billing_codes=AmiBillingCodes(name="Hourly2", codes=["bp-fake"]),
+            image_id="ami-fake-200-B",
+        ),
+    ]
+
+
+def test_pub_source_invalid_task_id(requests_mock, caplog):
+    """
+    Tests behavior when invalid task id is passed to source.
+    """
+    # test setup
+    caplog.set_level(logging.WARNING)
+    task_id = "a7df9"
+    pub_url = "https://pub.example.com"
+    request_url = os.path.join(
+        pub_url, "pub/task", "123456", "log/images.json?format=raw"
+    )
+
+    requests_mock.register_uri("GET", request_url, json=make_response(123456))
+
+    # request push items from source
+    with Source.get("pub:%s" % pub_url, task_id=task_id) as source:
+        push_items = [item for item in source]
+
+    # no exception raise but also no push items
+    assert len(push_items) == 0
+    # with follow line captured in log
+    assert caplog.messages[0] == "Invalid task_id: a7df9"
+
+
+def test_pub_source_empty_response(requests_mock, caplog):
```

Comment:
I don't think it should behave this way actually.  Can we please make it so that the absence of images.json is not considered an error?  If there are no AMIs I think it should just successfully produce nothing.

In general, source behavior is like this: get and return all items of all supported types found on the referenced object (staging directory; ET advisory; koji build).  It's not an error if any specific type has no content, or even if all types have no content.  I know "all supported types" for pub backend right now is only AMIs, but it still makes sense for the behavior to be consistent with other backends and to be consistent with future behavior of pub backend once other types are also supported.

If you're worried about users giving a wrong task ID and not noticing that there are no items produced, it could log a warning in the case where no items are found.